Extend `exclude` with `php` and `dependencies` settings

boesing opened the pull request:
<!--
Fill in the relevant information below to help triage your issue.

Pick the target branch based on the following criteria:
  * Documentation improvement: default X.Y.z branch or the oldest support X.Y.z
  * Bugfix: default X.Y.z branch or the oldest support X.Y.z
  * QA improvement (additional tests, CS fixes, etc.) that does not change code
    behavior: default X.Y.z branch or the oldest support X.Y.z
  * New feature, or refactor of existing code: develop branch

You MUST provide a signoff in your commits for us to be able to accept your
patch; you can do this by providing either the --signoff or -s flag when using
"git commit". Please see the project contributing guide and
https://developercertificate.org for details.
-->

|    Q          |   A
|-------------- | ------
| Documentation | yes
| New Feature   | yes

### Description

<!--
Tell us about why this change is necessary:
- Are you fixing a bug or providing a failing unit test to demonstrate a bug?
  - How do you reproduce it?
  - What did you expect to happen?
  - What actually happened?
  - TARGET THE default X.Y.z branch or the oldest support X.Y.z

- Are you adding documentation?
  - TARGET THE default X.Y.z branch or the oldest support X.Y.z

- Are you providing a QA improvement (additional tests, CS fixes, etc.) that
  does not change behavior?
  - Explain why the changes are necessary
  - TARGET THE default X.Y.z branch or the oldest support X.Y.z

- Are you fixing a BC Break?
  - How do you reproduce it?
  - What was the previous behavior?
  - What is the current behavior?
  - TARGET THE default X.Y.z branch or the oldest support X.Y.z

- Are you adding something the library currently does not support?
  - Why should it be added?
  - What will it enable?
  - How will the code be used?
  - TARGET THE develop BRANCH

- Are you refactoring code?
  - Why do you feel the refactor is necessary?
  - What types of refactoring are you doing?
  - TARGET THE develop BRANCH
-->

This introduces the ability to exclude specific jobs for specific PHP and/or composer dependency sets.
Fixes #93

boesing committed:
feature: allow to exclude specific PHP versions and/or dependency sets

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
docs: describe how job exclusion works with the most recent changes

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

froschdesign commented on README.md:
````diff
@@ -200,18 +200,45 @@ The tool discovers checks first, then appends any `additional_checks` are concat
+}
+```
+
+In case you only want to exclude jobs for specific composer dependency sets, add `dependencies` to the `exclude` configuration:
````
"Composer" is a proper name and for these names we respect the correct notation.

```suggestion
In case you only want to exclude jobs for specific Composer dependency sets, add `dependencies` to the `exclude` configuration:
```

froschdesign commented on README.md:
```diff
@@ -200,18 +200,45 @@ The tool discovers checks first, then appends any `additional_checks` are concat
 ### Excluding specific jobs
 
-The easiest way to exclude a single job is via the `name` parameter:
+You can exclude specific jobs by using their names:
```
For killing the word "easiest" 👍🏻 

Otherwise nice additions with examples to the different options.

boesing committed:
bugfix: use `job.tool.name` in case it is available

This fixes the problem that the job name does not match for tools due to the fact that the name already contains the human-readable name containing the version and dependencies string.

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: add debug logging on why a job might not be excluded

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

Ocramius commented:
@internalsystemerror is the CI failure related to what you fixed in #190 ?

internalsystemerror commented:
I believe you mean @gsteel although I see it losing the "before_script" from the matrix output for some reason.

I did some similar changes in https://github.com/laminas/laminas-ci-matrix-action/pull/167 and https://github.com/laminas/laminas-ci-matrix-action/pull/166 but upon reviewing those, I don't think they would affect this either.

internalsystemerror commented:
It looks like the only issue here is that it's no longer adding the `xmllint` checks in the matrix output whilst that is expected. Given the only recent change is #190, I suspect the top change (the one I commented on) needs reverting and adding `multiline: true` is all that's required.

boesing committed:
docs: use `Composer` brand name where possible

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

Ocramius commented on README.md:
```diff
@@ -200,18 +200,45 @@ The tool discovers checks first, then appends any `additional_checks` are concat
   "exclude": [
     {
-      "name": "PHPUnit on PHP 8.0 with latest dependencies"
+      "name": "PHPUnit"
```
Will the old exclusion via exact name match still work? I saw no changes in tests, which seems to hint at that, but I'm not 100% sure that's covered :D

boesing replied:
yes that would work but something is still not properly working and thus have to re-check here. No need to have this as a blocker for v2. I can adapt.

Ocramius replied:
Nono, I mean that you can merge and release `1.23.x` whenever: I'll deal with `2.0.x` over the weekend, if I can.

Ocramius reviewed (approved):
LGTM :+1: 

@boesing merge at your discretion after checking my last question :+1: 

We can also tag a new minor, before going to a new major with #189

Ocramius commented:
@boesing @internalsystemerror should I rebase and release this to `1.23.x`, if green?

boesing commented:
It is not properly working yet, please ignore this for any release. I'll handle this once I find some time to work on this. There is no pressure on this topic and no high need (at least for me) as of now 👍🏼

Ocramius commented:
Ok, just beware that it will go into `2.1.x`, probably.

boesing committed:
bugfix: we always have to check for the `job.name`

The exclusion strategy, when it was introduced, was meant to handle exact job names. By having the new feature of excluding tool names as well, we should still support the fill job.name as well so that we can still properly exclude jobs based on the initial exclusion strategy.

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
refactor: reduce the cognitive complexity of `app#isJobExcludedByConfiguration`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: add missing `test.env` files

With one of the newer versions of the matrix action, `test.env` files were introduced and thus have to be added to the new test cases as well.

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: put `php` schema definition into dedicated `definition`

This allows us to reference to that definition which eases the usage in several parts of the configuration. Having the list of supported PHP versions maintained in more than one location may become annoying and thus this changes makes a lot of sense.

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: add `before_script` section to the matrix output

This was introduced in a more recent matrix version and thus was missing the expected matrix from the initial version of this feature.

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: normalize PHP version for schema and also ensure that `stablePHP` from `.laminas-ci.json` reflects proper PHP version

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: target `stablePHP` to `installablePhpVersion` definition

Stable PHP should represent an actual PHP version and thus `lowest` and `latest` or even wildcard references make no sense.

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: use `allOf` to merge `installablePhpVersion` enum values with `php` enum values

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: use `anyOf` of `php` as just one of the `enum` values have to match, not `all`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: use `anyOf` to reference enum from `installablePhpVersion` and `php` specific enum strings

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: use `@default` for `no-checks`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: remove `locked` dependency from expected matrix

Due to the fact that there is no `composer.lock` file available, the matrix will not generate any `locked` jobs.

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>
